core/services/gateway/gateway.go

```diff
@@ -140,7 +138,17 @@ func (g *gateway) Close() error {
 }
 
 // Called by the server
-func (g *gateway) ProcessRequest(ctx context.Context, rawRequest []byte) (rawResponse []byte, httpStatusCode int) {
+func (g *gateway) ProcessRequest(ctx context.Context, rawRequest []byte, auth string) (rawResponse []byte, httpStatusCode int) {
+	jsonRequest, _ := g.jsonrpcHandler.DecodeRequest(rawRequest, auth)
+
+	if jsonRequest.Version == jsonrpc2.JsonRpcVersion {
+		return g.processJsonRpc2Request(ctx, jsonRequest, auth)
+	} else {
+		return g.processLegacyRequest(ctx, rawRequest)
```

Comment:
I think we need a different way to route. Legacy requests are also JSONRPC 2.0 requests. See [here](https://github.com/smartcontractkit/chainlink/blob/develop/core/scripts/gateway/web_api_trigger/invoke_trigger.go#L20) for a sample legacy request

Reply:
+1, this detection is not enough. What's wrong with the previous way? (decode + validate with the legacy codec).
Once you do that, you can do handler lookup based on wither donID or method name.